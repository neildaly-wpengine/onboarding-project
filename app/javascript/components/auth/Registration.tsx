import {
  Avatar,
  Button,
  Container,
  Grid,
  TextField,
  Typography,
} from "@material-ui/core";
import { makeStyles, Theme } from "@material-ui/core/styles";
import PersonIcon from "@material-ui/icons/Person";
import React, { useState } from "react";
import { Link, useHistory } from "react-router-dom";
import {
  ConsumerProps,
  NotifyAuthProps,
  Registration,
  RegistrationBody,
  RegistrationUser,
} from "../../common/types";

const useStyles = makeStyles((theme: Theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: "flex",
    flexDirection: "column",
    padding: 10,
    alignItems: "center",
  },
  avatar: {
    margin: theme.spacing(3),
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    width: "100%",
    marginTop: theme.spacing(3),
  },
  submit: {
    margin: theme.spacing(3, 0, 4),
  },
  link: {
    color: theme.palette.secondary.dark,
  },
}));

const Registration: React.FC<ConsumerProps & NotifyAuthProps> = ({
  consumer,
  toggleAuthentication,
}) => {
  const [confirmPasswordHelper, setConfirmPasswordHelper] = useState("");
  const [emailHelper, setEmailHelper] = useState("");
  const [registrationUser, setRegistrationUser] = useState<RegistrationUser>({
    email: "",
    firstName: "",
    lastName: "",
    passwordConfirmation: "",
    password: "",
  });
  const history = useHistory();

  const validPasswordEntries = (): boolean => {
    const valid: boolean =
      registrationUser?.password === registrationUser?.passwordConfirmation;

    valid ? setConfirmPasswordHelper("") : null;

    return valid;
  };

  const handleChange = (e: React.BaseSyntheticEvent): void => {
    const { name, value } = e.target;
    setRegistrationUser({ ...registrationUser, [name]: value });
  };

  const handleSubmit = (e: React.BaseSyntheticEvent): void => {
    e.preventDefault();

    if (!validPasswordEntries()) {
      setConfirmPasswordHelper("Passwords do not match.");
      return;
    }
    attemptUserRegistration();
  };

  const attemptUserRegistration = async () => {
    const registrationResponse = await consumer.registerNewUser({
      user: registrationUser,
    } as RegistrationBody);

    if (registrationResponse.data.status !== "created") {
      setEmailHelper(registrationResponse.data.message.email[0]);
      return;
    }
    toggleAuthentication();
    history.push("/");
  };

  const classes = useStyles();
  return (
    <Container component="main" maxWidth="xs">
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <PersonIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          Register
        </Typography>
        <form className={classes.form} onSubmit={handleSubmit}>
          <Grid container spacing={2}>
            <Grid item xs={12} sm={6}>
              <TextField
                autoComplete="fname"
                name="firstName"
                inputProps={{ "data-testid": "firstNameInput" }}
                onChange={handleChange}
                variant="outlined"
                required
                fullWidth
                id="firstName"
                label="First Name"
                autoFocus
              />
            </Grid>
            <Grid item xs={12} sm={6}>
              <TextField
                variant="outlined"
                required
                onChange={handleChange}
                inputProps={{ "data-testid": "lastNameInput" }}
                fullWidth
                id="lastName"
                label="Last Name"
                name="lastName"
                autoComplete="lname"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                variant="outlined"
                inputProps={{ "data-testid": "emailInput" }}
                required
                onChange={handleChange}
                fullWidth
                helperText={emailHelper}
                error={emailHelper !== ""}
                id="email"
                type="email"
                label="Email Address"
                name="email"
                autoComplete="email"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                variant="outlined"
                required
                onChange={handleChange}
                fullWidth
                name="password"
                label="Password"
                type="password"
                inputProps={{ "data-testid": "passwordInput" }}
                id="password"
                autoComplete="current-password"
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                variant="outlined"
                required
                onChange={handleChange}
                fullWidth
                name="passwordConfirmation"
                label="Confirm Password"
                type="password"
                inputProps={{ "data-testid": "passwordConfirmationInput" }}
                helperText={confirmPasswordHelper}
                error={confirmPasswordHelper !== ""}
                id="passwordConfirmation"
              />
            </Grid>
          </Grid>
          <Button
            data-testid="submit"
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            Register
          </Button>
          <Grid container justify="center">
            <Grid item>
              <Typography variant="subtitle1" gutterBottom>
                Already have an account?{" "}
                <Link to="/login" className={classes.link}>
                  Sign In.
                </Link>
              </Typography>
            </Grid>
          </Grid>
        </form>
      </div>
    </Container>
  );
};

export default Registration;
