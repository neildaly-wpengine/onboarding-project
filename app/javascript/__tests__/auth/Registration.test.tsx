import "@testing-library/jest-dom/extend-expect";
import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from "@testing-library/react";
import axios from "axios";
import humps from "humps";
import * as React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import Registration from "../../components/auth/Registration";
import { consumer } from "../articles/__helpers__/test-data";
import {
  registrationBody as body,
  successfulRegistrationResponse,
} from "./__helpers__/test-data";

const renderHelper = (): RenderResult => {
  return render(
    <Router>
      <Registration consumer={consumer} />;
    </Router>
  );
};

afterEach(cleanup);
jest.mock("axios");

describe("<Registration />", () => {
  describe("Snapshots", () => {
    test("should render an article highlight without crashing", () => {
      const { container } = renderHelper();

      expect(container).toMatchSnapshot();
    });
  });

  describe("Form inputs", () => {
    test("should update when changed", () => {
      const { getByTestId } = renderHelper();
      const emailInput = getByTestId("emailInput");
      const firstNameInput = getByTestId("firstNameInput");
      const lastNameInput = getByTestId("lastNameInput");

      fireEvent.change(emailInput, { target: { value: "test@test.com" } });
      fireEvent.change(firstNameInput, { target: { value: "Test" } });
      fireEvent.change(lastNameInput, { target: { value: "User" } });

      expect(emailInput).toHaveValue("test@test.com");
      expect(firstNameInput).toHaveValue("Test");
      expect(lastNameInput).toHaveValue("User");
    });
  });

  describe("Form submission", () => {
    test("should show error with unequal passwords", () => {
      const { getByTestId, container } = renderHelper();
      const passwordInput = getByTestId("passwordInput");
      const passwordConfirmationInput = getByTestId(
        "passwordConfirmationInput"
      );
      const submit = getByTestId("submit");

      fireEvent.change(passwordInput, {
        target: {
          value: body.user.password,
        },
      });
      fireEvent.change(passwordConfirmationInput, {
        target: {
          value: body.user.passwordConfirmation + "d",
        },
      });

      act(() => {
        fireEvent.click(submit);
      });

      expect(
        container.querySelector("#passwordConfirmation-helper-text")
      ).toBeInTheDocument();
      expect(
        container.querySelector("#passwordConfirmation-helper-text")
      ).toHaveTextContent("Passwords do not match.");
    });

    test("should register with adequate input", async () => {
      const { getByTestId } = renderHelper();
      const emailInput = getByTestId("emailInput");
      const firstNameInput = getByTestId("firstNameInput");
      const lastNameInput = getByTestId("lastNameInput");
      const passwordInput = getByTestId("passwordInput");
      const passwordConfirmationInput = getByTestId(
        "passwordConfirmationInput"
      );
      const expectedResponse = {
        data: successfulRegistrationResponse,
      };

      fireEvent.change(firstNameInput, {
        target: { value: body.user.firstName },
      });
      fireEvent.change(lastNameInput, {
        target: { value: body.user.lastName },
      });
      fireEvent.change(emailInput, {
        target: { value: body.user.email },
      });
      fireEvent.change(passwordInput, {
        target: { value: body.user.password },
      });
      fireEvent.change(passwordConfirmationInput, {
        target: { value: body.user.passwordConfirmation },
      });

      (axios.post as jest.Mock).mockImplementationOnce(() =>
        Promise.resolve(expectedResponse)
      );

      act(() => {
        fireEvent.click(getByTestId("submit"));
      });

      expect(axios.post).toBeCalledTimes(1);
      expect(axios.post).toBeCalledTimes(1);
      expect(axios.post).toHaveBeenCalledWith(
        "/api/v1/registrations",
        humps.decamelizeKeys(body),
        { withCredentials: true }
      );
    });
  });
});
