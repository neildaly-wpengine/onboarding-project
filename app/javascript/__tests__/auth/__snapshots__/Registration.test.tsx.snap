// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> Snapshots should render an article highlight without crashing 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <div
      class="makeStyles-paper-1"
    >
      <div
        class="MuiCollapse-container MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              class="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-alert-6 MuiPaper-elevation0"
              role="alert"
            >
              <div
                class="MuiAlert-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
              </div>
              <div
                class="MuiAlert-message"
              >
                That email is already in use!
              </div>
              <div
                class="MuiAlert-action"
              >
                <button
                  aria-label="close"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5"
      >
        Register
      </h1>
      <form
        class="makeStyles-form-3"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
                data-shrink="true"
                for="firstName"
                id="firstName-label"
              >
                First Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  autocomplete="fname"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="firstNameInput"
                  id="firstName"
                  name="firstName"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9 PrivateNotchedOutline-legendNotched-10"
                  >
                    <span>
                      First Name
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="lastName"
                id="lastName-label"
              >
                Last Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  autocomplete="lname"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="lastNameInput"
                  id="lastName"
                  name="lastName"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9"
                  >
                    <span>
                      Last Name
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email Address
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  autocomplete="email"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="emailInput"
                  id="email"
                  name="email"
                  required=""
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9"
                  >
                    <span>
                      Email Address
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  autocomplete="current-password"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="passwordInput"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9"
                  >
                    <span>
                      Password
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                data-shrink="false"
                for="passwordConfirmation"
                id="passwordConfirmation-label"
              >
                Confirm Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="passwordConfirmationInput"
                  id="passwordConfirmation"
                  name="passwordConfirmation"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9"
                  >
                    <span>
                      Confirm Password
                       *
                    </span>
                  </legend>
                </fieldset>
              </div>
              
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Register
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
            >
              Already have an account?
               
              <a
                class="makeStyles-link-5"
                href="/login"
              >
                Sign In.
              </a>
            </h6>
          </div>
        </div>
      </form>
    </div>
  </main>
  ;
</div>
`;
